### ISSUE TITLE
Should @storybook/react depend on `ajv`/`ajv-keywords`?

### ISSUE THREAD

**Issue by majapw** [storybooks/storybook#2979]
### Issue details
Hi all, 

I've been running through a pretty interesting dependency issue that I'm trying to figure out how best to mitigate. For context I am on:
npm@5.6.0
node@8.9.1

I have storybook as a dev dependency of my package as well as a whole lot of other things. Notably, storybook brings in webpack@^3.11.0 and postcss-loader@^2.1.0=>schema-utils@^0.4.4 ... both of which bring in ajv@6/ajv-keywords@3 as of a few days ago.

These changes down the tree made my app's `npm ls` (and therefore my tests) all start erroring out with no code changes. 

I have a basic repro case here: https://github.com/majapw/ajv-deps-test

Basically if you install `eslint` at the latest (which pulls in ajv@5) and `@storybook/react` at the latest (which pulls in ajv@6) you get an error when you run `npm ls` due to peer dep incompatibility. I *think* that if storybook depended directly on ajv@6, this would mitigated. I could be wrong though and am open to other possibilities. Thoughts?

@backwardok @ljharb @gabergg

### Steps to reproduce
```
git clone git@github.com:majapw/ajv-deps-test.git
npm install
npm ls
```

### Please specify which version of Storybook and optionally any affected addons that you're running
@storybook/react@3.3.13

### Affected platforms 
react

**Comment by Hypnosphi** [comment]
This part looks really weird, given that `6.1.1` matches `^6.0.0` mask:
```
UNMET PEER DEPENDENCY ajv@6.1.1
```

Iooks like an npm bug to me

**Comment by backwardok** [comment]
@Hypnosphi it's unmet because [eslint requires](https://github.com/eslint/eslint/blob/master/package.json#L37) `^5.3.0`, not because storybook requires `^6.0.0`

**Comment by majapw** [comment]
@Hypnosphi I think it's mostly that the error messages are pretty... unclear. But it seems to be that the lack of an `ajv` declaration at the top of that tree means there's some confusion as to what `ajv` package to use... and in this case, there are two.

**Comment by ljharb** [comment]
Specifically, if any of the direct deps [here](https://github.com/storybooks/storybook/blob/master/app/react/package.json) specifies a peer dependency that storybook/react does not explicitly declare as a dep, then that ever dep is an implicit, but required part of storybook/react’s API.

What happens if you cd into that package, npm install, and then run `npm ls`?

**Comment by Hypnosphi** [comment]
@backwardok why does it write `6.1.1` not `5.3.0` then

@ljharb I can't find any single package in the tree that depends on `ajv-keywords@3` but doesn't depend on its peer dep `ajv@6` at the same time

**Comment by Hypnosphi** [comment]
Probably relevant: https://github.com/npm/npm/issues/19483

**Comment by backwardok** [comment]
@Hypnosphi because it installed `5.3.0` and then didn't install `6.1.1`, which results in the unmet dependency of `^6.0.0`

**Comment by Hypnosphi** [comment]
> didn't install `6.1.1`

@backwardok in fact it did (at least for me with `npm@5.5.1`)

**Comment by Hypnosphi** [comment]
```
$ cat node_modules/webpack/node_modules/ajv/package.json
{
  "_from": "ajv@^6.1.0",
  "_id": "ajv@6.1.1",
  "_inBundle": false,
  "_integrity": "sha1-l41Zf7wrfQ5aXD3esUmmgvKr+g4=",
  "_location": "/webpack/ajv",
  "_phantomChildren": {},
  "_requested": {
    "type": "range",
    "registry": true,
    "raw": "ajv@^6.1.0",
    "name": "ajv",
    "escapedName": "ajv",
    "rawSpec": "^6.1.0",
    "saveSpec": null,
    "fetchSpec": "^6.1.0"
  },
  "_requiredBy": [
    "/webpack"
  ],
  "_resolved": "https://registry.npmjs.org/ajv/-/ajv-6.1.1.tgz",
  "_shasum": "978d597fbc2b7d0e5a5c3ddeb149a682f2abfa0e",
  "_spec": "ajv@^6.1.0",
  "_where": "/Users/jetbrains/ajv-deps-test/node_modules/webpack",
  "author": {
    "name": "Evgeny Poberezkin"
  },
  "bugs": {
    "url": "https://github.com/epoberezkin/ajv/issues"
  },
  "bundleDependencies": false,
  "dependencies": {
    "fast-deep-equal": "^1.0.0",
    "fast-json-stable-stringify": "^2.0.0",
    "json-schema-traverse": "^0.3.0"
  },
  "deprecated": false,
  "description": "Another JSON Schema Validator",
  "devDependencies": {
    "ajv-async": "^1.0.0",
    "bluebird": "^3.1.5",
    "brfs": "^1.4.3",
    "browserify": "^15.0.0",
    "chai": "^4.0.1",
    "coveralls": "^3.0.0",
    "del-cli": "^1.1.0",
    "dot": "^1.0.3",
    "eslint": "^4.14.0",
    "gh-pages-generator": "^0.2.0",
    "glob": "^7.0.0",
    "if-node-version": "^1.0.0",
    "js-beautify": "^1.7.3",
    "jshint": "^2.9.4",
    "json-schema-test": "^2.0.0",
    "karma": "^2.0.0",
    "karma-chrome-launcher": "^2.0.0",
    "karma-mocha": "^1.1.1",
    "karma-phantomjs-launcher": "^1.0.0",
    "karma-sauce-launcher": "^1.1.0",
    "mocha": "^5.0.0",
    "nyc": "^11.0.2",
    "phantomjs-prebuilt": "^2.1.4",
    "pre-commit": "^1.1.1",
    "require-globify": "^1.3.0",
    "typescript": "^2.6.2",
    "uglify-js": "^3.3.1",
    "watch": "^1.0.0"
  },
  "files": [
    "lib/",
    "dist/",
    "scripts/",
    "LICENSE",
    ".tonic_example.js"
  ],
  "homepage": "https://github.com/epoberezkin/ajv",
  "keywords": [
    "JSON",
    "schema",
    "validator",
    "validation",
    "jsonschema",
    "json-schema",
    "json-schema-validator",
    "json-schema-validation"
  ],
  "license": "MIT",
  "main": "lib/ajv.js",
  "name": "ajv",
  "nyc": {
    "exclude": [
      "**/spec/**",
      "node_modules"
    ],
    "reporter": [
      "lcov",
      "text-summary"
    ]
  },
  "repository": {
    "type": "git",
    "url": "git+https://github.com/epoberezkin/ajv.git"
  },
  "scripts": {
    "build": "del-cli lib/dotjs/*.js '!lib/dotjs/index.js' && node scripts/compile-dots.js",
    "bundle": "del-cli dist && node ./scripts/bundle.js . Ajv pure_getters",
    "bundle-beautify": "node ./scripts/bundle.js js-beautify",
    "eslint": "if-node-version \">=4\" eslint lib/*.js lib/compile/*.js spec/*.js scripts",
    "jshint": "jshint lib/*.js lib/**/*.js --exclude lib/dotjs/**/*",
    "prepublish": "npm run build && npm run bundle",
    "test": "npm run jshint && npm run eslint && npm run test-ts && npm run build && npm run test-cov && if-node-version 4 npm run test-browser",
    "test-browser": "del-cli .browser && npm run bundle && scripts/prepare-tests && npm run test-karma",
    "test-cov": "nyc npm run test-spec",
    "test-debug": "mocha spec/*.spec.js --debug-brk -R spec",
    "test-fast": "AJV_FAST_TEST=true npm run test-spec",
    "test-karma": "karma start --single-run --browsers PhantomJS",
    "test-spec": "mocha spec/*.spec.js -R spec $(if-node-version 7 echo --harmony-async-await)",
    "test-ts": "tsc --target ES5 --noImplicitAny lib/ajv.d.ts",
    "watch": "watch 'npm run build' ./lib/dot"
  },
  "tonicExampleFilename": ".tonic_example.js",
  "typings": "lib/ajv.d.ts",
  "version": "6.1.1"
}
```

**Comment by majapw** [comment]
Yeah it's weird because i seem to have both in my tree. 
```
➜  ajv-deps-test git:(master) cat node_modules/ajv/package.json| grep \"version\"
  "version": "5.5.2"
➜  ajv-deps-test git:(master) cat node_modules/webpack/node_modules/ajv/package.json | grep \"version\" 
  "version": "6.1.1"
```

**Comment by majapw** [comment]
My [basic example](https://github.com/majapw/ajv-deps-test) has many copies of ajv:
```
node_modules/webpack/node_modules/ajv/package.json
  "version": "6.1.1"
node_modules/uglifyjs-webpack-plugin/node_modules/ajv/package.json
 "version": "6.1.1"
node_modules/postcss-loader/node_modules/ajv/package.json
  "version": "6.1.1"
node_modules/ajv/package.json
"version": "5.5.2"
node_modules/fsevents/node_modules/ajv/package.json
"version": "4.11.8"
```

**Comment by majapw** [comment]
The bug is an npm bug! Woooo!

https://github.com/eslint/eslint/pull/10022#issuecomment-368964770

Closing this. :/
